pkg/private/zip/build_zip.py

```diff
@@ -68,53 +69,144 @@ def parse_date(ts):
-    dst_path += '/'
-  entry_info = zipfile.ZipInfo(filename=dst_path, date_time=ts)
-  # See http://www.pkware.com/documents/casestudies/APPNOTE.TXT
-  # denotes UTF-8 encoded file name.
-  entry_info.flag_bits |= 0x800
-  if mode:
-    f_mode = int(mode, 8)
-  else:
-    f_mode = default_mode
-
-  # See: https://trac.edgewall.org/attachment/ticket/8919/ZipDownload.patch
-  # external_attr is 4 bytes in size. The high order two bytes represent UNIX
-  # permission and file type bits, while the low order two contain MS-DOS FAT file
-  # attributes.
-  entry_info.external_attr = f_mode << 16
-  if entry_type == manifest.ENTRY_IS_FILE:
-    entry_info.compress_type = zipfile.ZIP_DEFLATED
-    with open(src, 'rb') as src:
-      zip_file.writestr(entry_info, src.read())
-  elif entry_type == manifest.ENTRY_IS_DIR:
-    entry_info.compress_type = zipfile.ZIP_STORED
-    # Set directory bits
-    entry_info.external_attr |= (UNIX_DIR_BIT << 16) | MSDOS_DIR_BIT
-    zip_file.writestr(entry_info, '')
-  elif entry_type == manifest.ENTRY_IS_LINK:
-    entry_info.compress_type = zipfile.ZIP_STORED
-    # Set directory bits
-    entry_info.external_attr |= (UNIX_SYMLINK_BIT << 16)
-    zip_file.writestr(entry_info, src)
-  # TODO(#309): All the rest
+
+class ZipWriter(object):
+
+  def __init__(self, output_path: str, time_stamp: int, default_mode: int):
+    """Create a writer.
+
+    You must close() after use or use in a 'with' statement.
+
+    Args:
+      output_path: path to write to
+      time_stamp: time stamp to add to files
+      default_mode: file mode to use if not specified in the entry.
+    """
+    self.output_path = output_path
+    self.time_stamp = time_stamp
+    self.default_mode = default_mode
+    self.zip_file = zipfile.ZipFile(self.output_path, mode='w')
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, t, v, traceback):
+    self.close()
+
+  def close(self):
+    self.zip_file.close()
+    self.zip_file = None
+
+  def make_zipinfo(self, path: str, mode: int):
+    """Create a Zipinfo.
+
+    Args:
+      path: file path
+      mode: file mode
+    """
+    entry_info = zipfile.ZipInfo(filename=path, date_time=self.time_stamp)
+    # See http://www.pkware.com/documents/casestudies/APPNOTE.TXT
+    # denotes UTF-8 encoded file name.
+    entry_info.flag_bits |= 0x800
+
+    # See: https://trac.edgewall.org/attachment/ticket/8919/ZipDownload.patch
+    # external_attr is 4 bytes in size. The high order two bytes represent UNIX
+    # permission and file type bits, while the low order two contain MS-DOS FAT
+    # file attributes.
+    if mode:
+      f_mode = int(mode, 8)
+    else:
+      f_mode = self.default_mode
+    entry_info.external_attr = f_mode << 16
+    return entry_info
+
+  def add_manifest_entry(self, options, entry):
+    """Add an entry to the zip file.
+
+    Args:
+      options: parsed options
+      zip_file: ZipFile to write to
+      entry: manifest entry
+    """
+
```

Comment:
Is there supposed to be an empty line after the docstrings like this?  This seems inconsistent with the preexisting functions in this file.

Reply:
Removed